tests/bluetooth/tester/src/btp_bas.c

```diff
@@ -0,0 +1,308 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stddef.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/services/bas.h>
+#include "btp/btp.h"
+#include "btp/btp_bas.h"
+
+static uint8_t bas_supported_commands(const void *cmd, uint16_t cmd_len, void *rsp,
+				      uint16_t *rsp_len)
+{
+	struct btp_bas_read_supported_commands_rp *rp = rsp;
+
+	/* octet 0 */
+	tester_set_bit(rp->data, BTP_BAS_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_BAS_SET_BATTERY_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_PRESENT);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRED_POWER_SOURCE);
+
+	/* octet 1 */
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_WIRELESS_POWER_SOURCE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_STATE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_LEVEL);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_CHARGE_TYPE);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_CHARGING_FAULT_REASON);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_IDENTIFIER);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_SERVICE_REQUIRED);
+	tester_set_bit(rp->data, BTP_BAS_BLS_SET_BATTERY_FAULT);
+
+	*rsp_len = sizeof(*rp) + 2;
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t btp_bas_set_battery_level(const void *cmd, uint16_t cmd_len, void *rsp,
+					 uint16_t *rsp_len)
+{
+	const struct btp_bas_set_battery_level_cmd *cp =
+		(const struct btp_bas_set_battery_level_cmd *)cmd;
+	uint8_t level = cp->level;
+	int err = bt_bas_set_battery_level(level);
+
+	return err ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+#if defined(CONFIG_BT_BAS_BLS)
+static uint8_t btp_bas_bls_set_battery_present(const void *cmd, uint16_t cmd_len, void *rsp,
+					       uint16_t *rsp_len)
+{
+	const struct btp_bas_bls_set_battery_present_cmd *cp =
+		(const struct btp_bas_bls_set_battery_present_cmd *)cmd;
+	enum bt_bas_bls_battery_present present = cp->present;
+
+	if (present != BT_BAS_BLS_BATTERY_NOT_PRESENT && present != BT_BAS_BLS_BATTERY_PRESENT) {
+		return BTP_STATUS_FAILED;
+	}
+
+	bt_bas_bls_set_battery_present(present);
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t btp_bas_bls_set_wired_power_source(const void *cmd, uint16_t cmd_len, void *rsp,
+						  uint16_t *rsp_len)
+{
+	const struct btp_bas_bls_set_wired_power_source_cmd *cp =
+		(const struct btp_bas_bls_set_wired_power_source_cmd *)cmd;
+	enum bt_bas_bls_wired_power_source source = cp->source;
+
+	if (source < BT_BAS_BLS_WIRED_POWER_NOT_CONNECTED || source > BT_BAS_BLS_WIRED_POWER_UNKNOWN) {
+		return BTP_STATUS_FAILED;
+	}
+
+	bt_bas_bls_set_wired_external_power_source(source);
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static uint8_t btp_bas_bls_set_wireless_power_source(const void *cmd, uint16_t cmd_len, void *rsp,
+						     uint16_t *rsp_len)
+{
+	const struct btp_bas_bls_set_wireless_power_source_cmd *cp =
+		(const struct btp_bas_bls_set_wireless_power_source_cmd *)cmd;
+	enum bt_bas_bls_wireless_power_source source = cp->source;
```

Comment:
I'm not 100% sure, but I think that @sjanc previously have disliked casting the BTP values to the enums like this, but not sure.

Reply:
ideally I'd prefer BTP defines and explicit assing, but if enum values are explicitly assigned and match BTP I'm fine with this, but is must be commented and source shall be validated to not be outside of enum values before assigning

Reply:
I moved the source verification before the assignment (and in all other applicable places in the file).

Regarding the commenting, I can see why it could be useful, but wouldn't it be redundant to have a comment in these functions when the actual btp_xx_cmd functions are just one search away?

Reply:
I guess the challenge is that the values in BTP are not really defined. 

So for example in `btp_bas_bls_set_wired_power_source` you do `enum bt_bas_bls_wired_power_source source = cp->source;`, but that assumes the the BTP value range and the Zephyr enum value range are identical. 

Zephyr (currently) follows the values defined by the BAS spec, but that's not something that is set in stone, and the values make change at any time (although unlikely). 

Ideally, and this is not specific for this PR or BAS, BTP would either define the value ranges itself (likely leaning on the value specified in the specs), and Zephyr would have BTP -> Zephyr conversion functions, similar to how we convert from and to HCI values from the Zephyr values like 
```c
uint8_t bt_get_phy(uint8_t hci_phy)
{
	switch (hci_phy) {
	case BT_HCI_LE_PHY_1M:
		return BT_GAP_LE_PHY_1M;
	case BT_HCI_LE_PHY_2M:
		return BT_GAP_LE_PHY_2M;
	case BT_HCI_LE_PHY_CODED:
		return BT_GAP_LE_PHY_CODED;
	default:
		return 0;
	}
}

int bt_get_df_cte_type(uint8_t hci_cte_type)
{
	switch (hci_cte_type) {
	case BT_HCI_LE_AOA_CTE:
		return BT_DF_CTE_TYPE_AOA;
	case BT_HCI_LE_AOD_CTE_1US:
		return BT_DF_CTE_TYPE_AOD_1US;
	case BT_HCI_LE_AOD_CTE_2US:
		return BT_DF_CTE_TYPE_AOD_2US;
	case BT_HCI_LE_NO_CTE:
		return BT_DF_CTE_TYPE_NONE;
	default:
		return BT_DF_CTE_TYPE_NONE;
	}
}
```

But that is way outside the scope of this PR

Reply:
Thanks for the great explanation. I added the BTP values range as a comment, and I will also update the doc file on the BTP (autopts) side to reflect this.